src/pad/src/ICeWall.cpp

```diff
@@ -0,0 +1,1070 @@
+  }
+}
+
+void ICeWall::removeFiller(odb::dbRow* row)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (row == nullptr) {
+    return;
+  }
+  const std::string prefix = fmt::format("IO_FILL_{}_", row->getName());
+
+  for (auto* inst : block->getInsts()) {
+    const std::string name = inst->getName();
+    if (name.substr(0, prefix.length()) == prefix) {
+      odb::dbInst::destroy(inst);
+    }
+  }
+}
+
+void ICeWall::placeBondPads(odb::dbMaster* bond,
+                            const std::vector<odb::dbInst*>& pads,
+                            odb::dbOrientType rotation,
+                            const odb::Point& offset,
+                            const std::string& prefix)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (bond == nullptr) {
+    return;
+  }
+
+  odb::dbMTerm* bond_pin = nullptr;
+  odb::Rect bond_rect;
+  odb::dbTechLayer* bond_layer = nullptr;
+  for (auto* mterm : bond->getMTerms()) {
+    for (auto* mpin : mterm->getMPins()) {
+      for (auto* geom : mpin->getGeometry()) {
+        auto* pin_layer = geom->getTechLayer();
+        if (pin_layer == nullptr) {
+          continue;
+        }
+        if (pin_layer->getRoutingLevel() == 0) {
+          continue;
+        }
+        bond_layer = pin_layer;
+        bond_pin = mterm;
+        bond_rect = geom->getBox();
+      }
+    }
+  }
+
+  const odb::dbTransform pad_xform(rotation);
+
+  for (auto* inst : pads) {
+    if (inst == nullptr) {
+      continue;
+    }
+    if (!inst->isFixed()) {
+      continue;
+    }
+
+    odb::dbTransform pad_transform(inst->getOrient());
+    odb::Point pad_offset = offset;
+    pad_transform.apply(pad_offset);
+    int x, y;
+    inst->getOrigin(x, y);
+    const odb::Point pad_loc(x + pad_offset.x(), y + pad_offset.y());
+
+    pad_transform.concat(pad_xform);
+    const odb::dbOrientType pad_orient = pad_transform.getOrient();
+
+    const std::string name = fmt::format("{}{}", prefix, inst->getName());
+
+    odb::dbInst* bond_inst = odb::dbInst::create(block, bond, name.c_str());
+    bond_inst->setOrient(pad_orient);
+    bond_inst->setOrigin(pad_loc.x(), pad_loc.y());
+    bond_inst->setPlacementStatus(odb::dbPlacementStatus::FIRM);
+
+    // connect bond and pad
+    odb::dbTransform xform;
+    bond_inst->getTransform(xform);
+    odb::Rect pin_shape = bond_rect;
+    xform.apply(pin_shape);
+    for (auto* iterm : inst->getITerms()) {
+      auto* net = iterm->getNet();
+      if (net == nullptr) {
+        continue;
+      }
+      auto* mterm = iterm->getMTerm();
+      for (auto* mpin : mterm->getMPins()) {
+        for (auto* geom : mpin->getGeometry()) {
+          auto* pin_layer = geom->getTechLayer();
+          if (pin_layer == bond_layer) {
+            bond_inst->getITerm(bond_pin)->connect(net);
+            makeBTerm(net, pin_layer, pin_shape);
+          }
+        }
+      }
+    }
+  }
+}
+
+void ICeWall::connectByAbutment()
```

Comment:
This is rather inefficient but perhaps the number of objects is small enough to get away with it.

Reply:
I haven't noticed a runtime issue yet, but I can probably look to see if there is a way to make it cleaner.